pilot/pkg/xds/workloadentry.go

```diff
@@ -0,0 +1,88 @@
+package xds
+
+import (
+	"fmt"
+	"istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/util"
+	"istio.io/istio/pkg/config"
+	"istio.io/istio/pkg/config/schema/gvk"
+)
+
+func (sg *InternalGen) RegisterWorkload(proxy *model.Proxy) {
+	if proxy.Metadata.AutoRegisterGroup == "" {
+		return
+	}
+	if len(proxy.IPAddresses) == 0 {
+		adsLog.Errorf("auto registration of %v failed: missing IP addresses", proxy.ID)
+		return
+	}
+	if len(proxy.Metadata.Namespace) == 0 {
+		adsLog.Errorf("auto registration of %v failed: missing namespace", proxy.ID)
+		return
+	}
+
+	// check if the WE already exists, update the status
+	entryName := autoregisteredWorkloadEntryName(proxy)
+	entryCfg := sg.Store.Get(gvk.WorkloadEntry, entryName, proxy.Metadata.Namespace)
+	if entryCfg != nil {
+		// TODO put this in status/use managed fields
+		entryCfg.Meta.Labels["workload-controller"] = sg.Server.instanceID
+		_, err := sg.Store.Update(*entryCfg)
+		if err != nil {
+			adsLog.Warn("failed to update auto-registered WorkloadEntry %s/%s: %v", proxy.Metadata.Namespace, entryName, err)
+		}
+		return
+	}
+
+	// create WorkloadEntry from WorkloadGroup
+	groupCfg := sg.Store.Get(gvk.WorkloadGroup, proxy.Metadata.AutoRegisterGroup, proxy.Metadata.Namespace)
+	if groupCfg == nil {
+		adsLog.Warnf("auto registration of %v failed: cannot find WorkloadGroup %s/%s", proxy.ID, proxy.Metadata.Namespace, proxy.Metadata.AutoRegisterGroup)
+		return
+	}
+	group := groupCfg.Spec.(*v1alpha3.WorkloadGroup)
```

Comment:
is the groupCfg we get back a copy? I forget. I think it is, but if not we probably need to copy it

Reply:
Looks like we copy the meta values but not the spec.

Reply:
the underlying istio/client-go might also do it. Probably not, but possible